Allow the scheduler to handle yielding a list of coroutines and triggers

TC01 opened the pull request:
Currently the cocotb scheduler understands how to yield a coroutine from another coroutine; it queues it and calls .join() automatically. It also understands how to yield a list of triggers and resume execution when the first one finishes.

But it does not understand how to yield a list of triggers and coroutines (or just coroutines); it will complain if the list contains anything that's not a trigger.

This commit fixes it so that if the scheduler encounters a list, it will iterate through that list and build a new one, calling ```.join()``` automatically on any objects of type RunningCoroutine. If the list contains something that was not a trigger or coroutine, it is dropped.

Then the size of the new and old lists are compared; if they differ it means that the list contained something that was not a trigger or a coroutine, and so we do nothing. Otherwise we yield it.

The commit also includes a very minimal test for this functionality.

This resolves issue #588.

TC01 commented:
Unfortunately it just occurred to me that this approach does not quite work.

If the processed and original lists are not equal in size (e.g. there was something that was not a trigger or a coroutine), the code won't call ```self._coroutine_yielded``` but it also won't branch into the error condition.

I guess this could be fixed by adding the list comprehension (```elif ... and not [t for t in result if not isinstance(t, Trigger)])```) back to the conditional but also looking for coroutines. I'll update this PR tomorrow with a fix.

eric-wieser commented:
Might it be better to add `yield cocotb.utils.All(one, two, three, ...)` rather than adding special cases for lists?

TC01 commented:
Perhaps, though maybe it should be a special trigger, kind of like the [Combine trigger](https://github.com/potentialventures/cocotb/blob/master/cocotb/triggers.py#L328)-- which now that I think of it, also needs extending to support coroutines.

The special case to yield a list of triggers already exists, though-- this specific PR is just about extending that to support coroutines too.

Anyway, I realize I never explicitly stated this, but: the second commit I pushed should resolve the issue I mentioned above. So this PR should be ready for review.

imphil commented:
@TC01 thanks for this. Would you mind rebasing this PR on top of the current master to resolve the merge conflicts? Also it would be nice if you could add documentation describing this functionality, so people know what's expected to work (and what is not).

TC01 commented:
Sure, I've rebased. I also added a note to the documentation that it is possible to yield a list of triggers and coroutines.

I have not updated the Combine trigger with similar functionality; I'll open another PR soon to do that though.

imphil commented:
Thanks!
